### PR TITLE
Feat 1397 - Implement Invoice object 

### PR DESCRIPTION
**Pull request by phpgeekfr**
Implement Invoice object

### CODE REVIEW

File: libs/contract/src/lib/contract/+state/contract.firestore.ts [hunk 1]
```diff
@@ -71,11 +72,20 @@ interface ContractRaw<D> {
   documents: LegalDocuments
 }
 
-/*
 export interface InvoiceRaw<D> {
-   @todo #1397
+  id: string,
+  internalRef: string,
+  paymentRef?: string, // @dev should be comming from blockchain data
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/contract/+state/contract.firestore.ts]: If you do it like that (I think) we cannot see it on hover. I would recommend to use JSDoc syntax (especially when it describe the data): 
```typescript
internalRef: string,
/** @dev should be comming from blockchain data */
paymentRef?: string,
```

File: libs/movie/src/lib/movie/components/import/table-extracted-deals/table-extracted-deals.component.ts [hunk 2]
```diff
@@ -124,7 +124,7 @@ export class TableExtractedDealsComponent implements OnInit {
     switch (type) {
       case 'start':
         if (!term.start || isNaN(term.start.getTime())) {
-          return term.startLag ? term.startLag : noDate;
+          return term.approxStart ? term.approxStart : noDate;
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/components/import/table-extracted-deals/table-extracted-deals.component.ts]: You can use the `||` operator for this kind of operation : `return term.approxStart || noDate`

File: libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts [hunk 3]
```diff
@@ -1320,30 +1320,19 @@ export class ViewExtractedElementsComponent {
...
-                      name: 'CatchUp start',
-                      reason: `Failed to parse CatchUp start date, moved data to startLag`,
-                      hint: 'Edit corresponding sheet field.'
-                    });
-                  } else {
-                    distributionDeal.catchUp.start = catchUpStartDate;
-                  }
-                } catch (error) {
-                  distributionDeal.catchUp.startLag = spreadSheetRow[SpreadSheetDistributionDeal.catchUpStartDate];
+                const ssfCatchUpStartDate: SSF$Date = SSF.parse_date_code(spreadSheetRow[SpreadSheetDistributionDeal.catchUpStartDate]);
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts]: javascript uses the `camelCase` convention for variables, keys, function and method name. I would recommend to change the method to `parseDateCode`

File: libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts [hunk 4]
```diff
@@ -1320,30 +1320,19 @@ export class ViewExtractedElementsComponent {
...
+                const catchUpStartDate = new Date(`${ssfCatchUpStartDate.y}-${ssfCatchUpStartDate.m}-${ssfCatchUpStartDate.d}`);
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts]: For better readability I would recommend : 
```typescript
const { y, m, d } = ssfCatchUpStartDate;
const catchUpStartDate = new Date(`${y}-${m}-${d}`);
```

File: libs/utils/src/lib/common-interfaces/price.ts [hunk 5]
```diff
@@ -30,10 +46,23 @@ export function createPrice(price: Partial<Price> = {}): Price {
 /**
  * A factory function that creates Fee
  */
-export function createFee(fee: Partial<Fee> = {}): Fee {
+export function createFee(params: Partial<Fee> = {}): Fee {
   return {
     label: '',
-    price: createPrice(fee.price),
-    ...fee
+    price: createPrice(params.price),
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/common-interfaces/price.ts]: Do you want it to be erased by the price coming from params if any ?

File: libs/utils/src/lib/common-interfaces/schedule.ts [hunk 6]
```diff
@@ -1,22 +1,23 @@
...
 
-
 export interface PaymentSchedule extends PaymentScheduleRaw<Date> {
 }
 
-
 export function createPaymentSchedule(params: Partial<PaymentSchedule> = {}): PaymentSchedule {
   return {
     percentage: 0,
+    date: createScheduledDate(params ? params.date : undefined),
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/common-interfaces/schedule.ts]: you can do `createScheduledDate(params && params.date)` and soon (angular 9) `params?date`

File: libs/utils/src/lib/common-interfaces/terms.ts [hunk 7]
```diff
@@ -33,11 +49,33 @@ export interface TermsRaw<D> {
...
+  period: FloatingDuration, // @dev this replace floatingDuration for better readability
+}
+
 export interface Terms extends TermsRaw<Date> {
 }
 
+export interface ScheduledDate extends ScheduledDateRaw<Date> {
+}
+
+export function createFloatingDuration(params: Partial<FloatingDuration> = {}): FloatingDuration {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/common-interfaces/terms.ts]: You don't want any default values ?

File: libs/utils/src/lib/common-interfaces/terms.ts [hunk 8]
```diff
@@ -33,11 +49,33 @@ export interface TermsRaw<D> {
...
+  return {
+    ...params
+  };
+}
+
 export function createTerms(params: Partial<Terms> = {}): Terms {
   return {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/common-interfaces/terms.ts]: Same here ?

File: libs/utils/src/lib/common-interfaces/utility.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    phoneNumber: '',
+    region: '',
+    ...params
+  };
+}
+
+/** A factory function that creates a BankAccount */
+export function createBankAccount(params: Partial<BankAccount> = {}): BankAccount {
+  return {
+    address: createLocation(params ? params.address : undefined),
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/common-interfaces/utility.ts]: Same as above

File: libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts (hunk 3 above)
Comment (phpgeekfr) [reply libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts]: This function is from node_modules directory. Libs handling excels have always been awful to look at

File: libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts (hunk 4 above)
Comment (phpgeekfr) [reply libs/movie/src/lib/movie/components/import/view-extracted-elements/view-extracted-elements.component.ts]: good one !

File: libs/utils/src/lib/common-interfaces/schedule.ts (hunk 6 above)
Comment (phpgeekfr) [reply libs/utils/src/lib/common-interfaces/schedule.ts]: This new form `params?date` will be a good time saving

File: libs/utils/src/lib/common-interfaces/terms.ts (hunk 7 above)
Comment (phpgeekfr) [reply libs/utils/src/lib/common-interfaces/terms.ts]: every `FloatingDuration` attribs are optionnals

File: libs/utils/src/lib/common-interfaces/terms.ts (hunk 8 above)
Comment (phpgeekfr) [reply libs/utils/src/lib/common-interfaces/terms.ts]: every `Terms` attribs are optional

**Pull request merged**